File: src/main/java/abstract_factory/creator/SuperVehicleFactory.java
package abstract_factory.creator;

import abstract_factory.product.SuperBus;
import abstract_factory.product.SuperTruck;
import abstract_factory.product.Vehicle;
import abstract_factory.product.VehicleType;
import abstract_factory.product.part.CityVehiclePartFactory;
import abstract_factory.product.part.OffloadVehiclePartFactory;

public class SuperVehicleFactory extends VehicleFactory {
    @Override
    protected Vehicle createVehicle(VehicleType vehicleType) {
        return switch (vehicleType) {
            case BUS -> new SuperBus(new CityVehiclePartFactory());
            case TRUCK -> new SuperTruck(new OffloadVehiclePartFactory());
        };
    }
}

File: src/main/kotlin/abstract_factory/creator/SuperVehicleFactory.kt
package abstract_factory.creator

import abstract_factory.product.SuperBus
import abstract_factory.product.SuperTruck
import abstract_factory.product.Vehicle
import abstract_factory.product.VehicleType
import abstract_factory.product.part.CityVehiclePartFactory
import abstract_factory.product.part.OffloadVehiclePartFactory

class SuperVehicleFactory : VehicleFactory() {
    override fun createVehicle(vehicleType: VehicleType): Vehicle {
        return when (vehicleType) {
            VehicleType.BUS -> SuperBus(CityVehiclePartFactory())
            VehicleType.TRUCK -> SuperTruck(OffloadVehiclePartFactory())
        }
    }
}